examples/toy/Ch4/mlir/ToyCombine.td

```diff
@@ -0,0 +1,26 @@
+
+#ifdef TOY_COMBINE
+#else
+#define TOY_COMBINE
+
+#ifdef OP_BASE
+#else
+include "mlir/IR/OpBase.td"
+include "include/toy/Ops.td"
+#endif // OP_BASE
+
+// Transpose(Transpose(x)) = x
+def TransposeOptPattern : Pat<(TransposeOp(TransposeOp $arg)), (replaceWithValue $arg)>;
+
+// Reshape(Reshape(x)) = x
+def ReshapeReshapeOptPattern : Pat<(ReshapeOp(ReshapeOp $arg)), (ReshapeOp $arg)>;
+
+// Reshape(Constant(x)) = x'
+def ReshapeConstant : NativeCodeCall<"reshapeConstant($_builder, $0)">;
```

Comment:
We *could* define the whole thing inline:

 def ReshapeConstant :
   NativeCodeCall<"$0.reshape($1->getType().cast<ShapedType>())">;

?

Reply:
That does look cleaner. Do you think it is still useful to have a separate call to the helper function to show how it can be used if required?

Reply:
It probably isn't necessary. As long as you properly describe that the native code call is actually inline c++, it should be fine IMO.

Reply:
I've updated the code to define the optimization inline.